hubblestack/hec/opt.py

```diff
@@ -30,17 +30,38 @@ class Required(object):
 del Required
 
 MODALITIES = ('grains.get','config.get',) # search in grains first, fallback to config.get
+options_for_grains_config = {'token', 'index', 'port'}
 
 def _get_splunk_options(space, modality, **kw):
     ret = list()
 
     confg = __salt__['config.get']
 
+    # both index and token must be specified if at all overriding in /etc/hubble/hubble
+    # is taking place using the variables splunk_token and splunk_index
+
+    if bool(confg('splunk_token', None)) != bool(confg('splunk_index', None)):
+        raise Exception('Both index and token must be specified together or not '
+                        'specified at all in case of overriding')
+
+    if confg('splunk_token', None) and not str(confg('splunk_token', None)).strip():
+        raise Exception('splunk_token cannot be an empty field')
+
+    if confg('splunk_index', None) and not str(confg('splunk_index', None)).strip():
+        raise Exception('splunk_index cannot be an empty field')
+
+    if confg('splunk_port', None) and not str(confg('splunk_port', None)).strip():
```

Comment:
You should also check if 'splunk_port' is integer value. It should not be a random string.

Reply:
I'm not sure checking integer or even str(config()) is necessary, and certainly not strip().

The salt yaml loader does most of that heavy lifting for us (str vs int and strip()). And I think if someone puts "dog" for the port number, it won't be a very different error than if they just use the wrong port. You'd get a network stack error and think: hrm, 'dog' might not be the right port number.

Reply:
(Also, if you were to put "splunk 8080" at the end of /etc/protocols, then "splunk" might just be a valid port number.)

Reply:
That's good info that salt's loader would take care of empty strings, didn't know that.
Regarding /etc/protocols entry, that would be a valid scenario not sure if someone would ever use it with Hubble.

For integer check I was thinking of handling typos in config (especially hubble.d/) w.r.t. port field. Salt wouldn't know if that variable should have integer or string values which we could handle and fail fast instead of giving weird connection errors. 

For most of Adobe's use-cases/deployments it wouldn't matter, it's more of 'good to have' sort of check.

Reply:
I've used text based port numbers plenty... both with things like socket.open('hostname:portname') and in config files. Some programs can deal with it and others can't because people check for digits.

I think it's fine to check that the port number is really a number here, so I won't make any further argument about it, but it'd make me smirk if I was an SRE and noticed it was so inflexible.

Futher notes on the yaml conversions... I think it wouldn't take care of empty strings if you told it not to do so. Definitely the salt code avoids checking types and `strip()`s in most places though. I think it'd make sense to stay consistent with the rest of salt and hubble and assume it's stripped.

But that said, if you used quotes to force the leading spaces, they'd survive into the `__opts__`.

```
blah:
   intentionally_spacey: '   leading spaces, and trailing ones   '
   autostripped:          this won't have leading spaces regardless, just part of yaml
```

Also, on the number conversions, when converting from yaml, salt is going to automatically convert all-number fields to int and leave strings alone... which is the source of some common errors in salt, particularly around octal.

```
config:
   some_integer: 755
   wrong_file_mode: 0755
   right_file_mode: '0755'
```

When a field is all number characters, the salt-yaml conversion makes it `int(755)`, so some_integer is fine... it's `int(755)` automatically. 

wrong_file_mode will actually give `int(755)==int(0o1363)` (sometimes) though. That's probably not what the salt user wanted. The salt-yaml conversion seems to not know about leading 0's in some implementations, does know about it in others. The salt best practices (and file.manged notes) therefore indicate that you should mark file modes as strings so they're handed to `chmod` correctly. That means using the quote markers to avoid the auto conversions: `'0755'`.

Reply:
ok. Let's not check for integer port numbers then.